File: Python/project_euler/problem_03.py
"""https://projecteuler.net/problem=3"""

from math import ceil, sqrt
from project_euler.utilities.primality import is_prime


def solve(integer: int = 600851475143) -> int:
    """Solves problem 3.

    Finds the largest prime of the argument
    through trial division by checking
    if the factor is a prime. The factors are
    checked in reverse order and starts with
    the square root of the integer.

    Arguments:
        integer {int} -- The integer to be factored.
        Default is 600851475143

    Returns:
        int -- The largest prime factor of the integer if found.
    """
    prime = None
    upper_factor = ceil(sqrt(integer))
    for i in range(upper_factor, 0, -1):
        if integer % i == 0 and is_prime(i):
            prime = i
            break
    return prime

File: CSharp/Miller.ProjectEuler.Solutions/Problem03.cs
using Miller.ProjectEuler.Solutions.Extensions;
using System;

namespace Miller.ProjectEuler.Solutions
{
    /// <summary>
    /// https://projecteuler.net/problem=3
    /// </summary>
    public sealed class Problem03 : IProblem
    {
        public string Name => "Largest prime factor";

        private ulong _number;

        public Problem03(ulong number = 600_851_475_143)
        {
            _number = number;
        }

        /// <summary>
        /// Solves problem 3 by testing each factor of the passed in number for primality
        /// and returns the largest one found. The numbers are tested in reverse
        /// order beginning with the square root of the passed in number.
        /// </summary>
        /// <returns>Largest prime factor</returns>
        public string Solve()
        {
            ulong limit = (ulong)Math.Ceiling(Math.Sqrt(_number));
            for (ulong i = limit; i > 0; i--)
            {
                if (_number % i == 0 && i.IsPrime())
                {
                    return i.ToString();
                }
            }
            throw new Exception($"No primes were found for number {_number}.");
        }
    }
}